source/adios2/toolkit/remote/Remote.cpp

```diff
@@ -50,16 +50,22 @@ void ReadResponseHandler(CManager cm, CMConnection conn, void *vevent, void *cli
     return;
 };
 
+CManagerSingleton &CManagerSingleton::Instance(RemoteCommon::Remote_evpath_state &ev_state)
+{
+    static CManagerSingleton self = [&ev_state] {
+        CManagerSingleton instance;
+        ev_state = instance.internalEvState;
+        return instance;
+    }();
+    ev_state = self.internalEvState;
```

Comment:
I am not sure but I think a thread might reach this line before the thread calling the lambda finished in which case there will be a thread with an invalid internal ev state. If this is the thread that ends up calling std::call_once we will have a problem. Is there a reason why you want the `CMregister_handler` calls outside the singleton?

Reply:
The state should be complete once we have registered the formats.  We don't need the handlers to have been registered until we get messages arriving.  (I had them outside of the Singleton because they weren't visible inside it.  I could have added the externs before it, but I don't think it should be necessary.). My understanding (which may be flawed), is that the singleton creation is thread safe, so anyone that gets it will get it after its constructor has been completed.  So, the formats should have been registered and the internalEvState fully populated.  So, the ev_state in every Remote instance should be appropriately populated.  One of them will register the handlers, and from what I read of the call_once docs, there will be no race conditions.  I.E. if a thread invokes the call_once, it will only proceed past that once someone has completed the call, even if it's some other thread doing it.  I don't think there are holes here, but I can't explain the failures.  Will be trying to get it into a docker image to see if I can get more info.

Reply:
Well, it wasn't the internal state thing.  Instead icc seems not agree that this pattern produces a singleton.  Specifically, the constructor gets called (creating the CManager), then the destructor gets called (closing the CManager), then we start registering handlers but get a segfault because the CManager has been closed and deallocated.   I'm guessing that the other failure, from nvhpc is from the same problem.  The gcc-based compilers seem happy, but not the rest.

Reply:
Yes you are right, the internal state is guaranteed to be created when threads reach that line. Let me play with godbolt a bit to look at the assembly created by icc. I am very surprised only gcc likes this.